onnxruntime/core/providers/cpu/tensor/scatter.cc

```diff
@@ -241,17 +285,29 @@ Status Scatter::Compute(OpKernelContext* context) const {
     }
   }
 
-  auto* data_output = context->Output(0, input_data_shape);
+  Status status{};
+  const auto index_type = indices_input->GetElementType();
+  std::vector<int64_t> indices_data{};
 
-  MLDataType Tdata_type = data_input->DataType();
-  Status status;
-  if (indices_input->IsDataType<int32_t>()) {
-    DispatchOnTensorTypeWithReturn(Tdata_type, status, CopyInt32Index, data_input, indices_input, updates_input, axis, data_output);
-  } else if (indices_input->IsDataType<int64_t>()) {
-    DispatchOnTensorTypeWithReturn(Tdata_type, status, CopyInt64Index, data_input, indices_input, updates_input, axis, data_output);
+  if (index_type == utils::ToTensorProtoElementType<int32_t>()) {
+    status = GetIndices<int32_t>(*data_input, *indices_input, axis, indices_data);
+  } else if (index_type == utils::ToTensorProtoElementType<int64_t>()) {
+    status = GetIndices<int64_t>(*data_input, *indices_input, axis, indices_data);
   } else {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Expecting indices to be either int32_t or int64_t");
+    status = ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Indices type is not supported.");
   }
+
+  if (!status.IsOK()) {
+    return status;
+  }
+
```

Comment:
nit: Can we move all the code prior to here into a non-templatized method so it's not duplicated?